File: packages/client-typescript/src/client/evals.ts
/** Managed evaluations v1 over bearer HTTP. No WebSocket, retries, or redirects. */
export type EvaluationVerdict = 'pass' | 'fail' | 'error' | 'incomplete' | 'abstain';
export interface EvaluationSpec {
	schemaVersion: 1;
	name: string;
	projectId: string;
	pipeline: Record<string, unknown>;
	source: string;
	inputMode: 'chat' | 'text';
	environment: 'development' | 'staging';
	datasetName: string;
	cases: Array<{ id: string; name: string; input: string; reference?: string; approved: boolean; tags?: string[]; provenance?: Record<string, unknown> }>;
	scorers: Array<{ id: string; name: string; kind: 'equals' | 'contains' | 'not_contains' | 'json' | 'latency' | 'llm_judge' | 'human'; expected?: string; path?: string; threshold?: number; rubric?: string; judgePipeline?: Record<string, unknown> }>;
	repetitions: number;
	passCriteria: { minimumPassRate: number; maxRegressions: number };
}
export interface ManagedEvaluation {
	id: string;
	name: string;
	projectId: string;
	revision: number;
	spec: EvaluationSpec;
	baselineRunId?: string;
	createdAt: string;
	updatedAt: string;
}
export interface EvaluationComparison {
	compatible: boolean;
	reasons: string[];
	regressions: number;
	improvements: number;
	unchanged: number;
	cases: Array<{ caseId: string; trial: number; baselineStatus: string; candidateStatus: string; change: string }>;
	pipelineChanges: unknown[];
}
export interface EvaluationRun {
	id: string;
	evaluationId: string;
	revision: number;
	name: string;
	status: 'queued' | 'running' | 'completed' | 'cancelled' | 'error';
	createdAt: string;
	startedAt?: string;
	finishedAt?: string;
	spec: EvaluationSpec;
	pipelineHash: string;
	datasetHash: string;
	scorerHash: string;
	baselineRunId?: string;
	baselineReportRevision?: number;
	reportRevision: number;
	summary: { total: number; completed: number; pass: number; fail: number; error: number; incomplete: number; abstain: number; passRate: number; durationMs: number; gate: 'pass' | 'fail' | 'incomplete'; costUsd: number | null };
	cases: Array<{
		id: string;
		caseId: string;
		name: string;
		trial: number;
		input: string;
		reference?: string;
		output?: string;
		status: EvaluationVerdict | 'pending' | 'running';
		durationMs: number;
		executionProjectId?: string;
		scores: Array<{ scorerId: string; name: string; status: EvaluationVerdict; score: number | null; reason: string; executionProjectId?: string; trace?: EvaluationTrace }>;
		trace?: EvaluationTrace;
		error?: string;
	}>;
	comparison?: EvaluationComparison;
	error?: string;
}
export interface EvaluationTrace {
	projectId: string;
	source: string;
	runKind?: 'dev' | 'deploy';
	environment?: 'development' | 'staging';
	traceId?: number;
	chapter?: number;
}
export interface EvaluationRunOptions {
	revision: number;
	idempotencyKey: string;
	baselineRunId?: string;
}
export interface EvaluationReview {
	caseId: string;
	caseResultId?: string;
	scorerId: string;
	status: 'pass' | 'fail' | 'abstain';
	reason: string;
	expectedReportRevision: number;
}
export interface EvaluationWaitOptions {
	timeout?: number;
	pollInterval?: number;
}
export interface EvaluationCapabilities {
	environments: Array<{ id: string; label: string; available: boolean }>;
	scorerKinds: string[];
	assistantAvailable: boolean;
}
export type EvaluationRunResponse = { run: EvaluationRun };

/** Safe failure metadata. Transport errors never include URLs or credentials. */
export class EvalsError extends Error {
	constructor(
		message: string,
		readonly status?: number,
		readonly code = 'evals_error'
	) {
		super(message);
		this.name = 'EvalsError';
	}
}

export function redactEvalsText(text: string, credential: string): string {
	return (credential ? text.split(credential).join('[REDACTED]') : text).replace(/Bearer\s+[^\s"<>]+/gi, 'Bearer [REDACTED]');
}

function baseUrl(uri: string): string {
	try {
		if (!uri || /[\s\\]/.test(uri) || [...uri].some((character) => character.charCodeAt(0) < 32) || uri.includes('?') || uri.includes('#')) throw new Error();
		const url = new URL(uri.includes('://') ? uri : `http://${uri}`);
		if (!['http:', 'https:', 'ws:', 'wss:'].includes(url.protocol) || !url.hostname || url.username || url.password || url.port === '0' || !['', '/task/service', '/evals/v1'].includes(url.pathname.replace(/\/+$/, ''))) throw new Error();
		url.protocol = url.protocol === 'wss:' ? 'https:' : url.protocol === 'ws:' ? 'http:' : url.protocol;
		return `${url.origin}/evals/v1`;
	} catch {
		throw new Error('Managed evaluations require an HTTP(S) server origin without credentials, query, or fragment');
	}
}

function id(value: string): string {
	if (typeof value !== 'string' || !value.trim() || value === '.' || value === '..') throw new Error('A non-empty object ID is required');
	return encodeURIComponent(value);
}
function revision(value: number, name = 'revision'): void {
	if (!Number.isSafeInteger(value) || value < 1) throw new Error(`${name} must be a positive integer`);
}
function positive(value: number, name: string): void {
	if (!Number.isFinite(value) || value <= 0 || value > 2_147_483_647) throw new Error(`${name} must be a positive finite number within the timer limit`);
}

/** ``client.evals``. Request/wait budgets and poll intervals are milliseconds. */
export class EvalsApi {
	constructor(
		private readonly connection: () => { uri: string; auth: string },
		readonly timeout = 30_000
	) {
		positive(timeout, 'timeout');
	}

	private async request<T>(method: string, path: string, body?: unknown, raw = false, timeout = this.timeout): Promise<T> {
		const { uri, auth } = this.connection();
		const url = baseUrl(uri) + path;
		if (!auth || /\s/.test(auth) || [...auth].some((character) => character.charCodeAt(0) < 32)) throw new Error('A bearer credential is required for managed evaluations');
		positive(timeout, 'timeout');
		const headers: Record<string, string> = { Authorization: `Bearer ${auth}`, Accept: raw ? 'application/xml' : 'application/json' };
		let payload: string | undefined;
		if (body !== undefined) {
			headers['Content-Type'] = 'application/json';
			try {
				payload = JSON.stringify(body, (_key, value) => {
					if (typeof value === 'number' && !Number.isFinite(value)) throw new Error();
					return value;
				});
			} catch {
				throw new Error('Request body must be strict JSON');
			}
		}
		const controller = new AbortController();
		const timer = setTimeout(() => controller.abort(), Math.min(timeout, this.timeout));
		let response: Response;
		let text: string;
		try {
			response = await fetch(url, { method, headers, body: payload, redirect: 'error', credentials: 'omit', signal: controller.signal });
			text = await response.text();
		} catch {
			throw new EvalsError(controller.signal.aborted ? 'Managed evaluation request timed out; no request was retried' : 'Managed evaluation request failed; no request was retried', undefined, controller.signal.aborted ? 'timeout' : 'transport_error');
		} finally {
			clearTimeout(timer);
		}
		if (!response.ok) {
			let message = `Managed evaluation request failed (HTTP ${response.status})`;
			let code = 'http_error';
			try {
				const error = JSON.parse(text).error;
				if (typeof error?.message === 'string') message = error.message;
				if (typeof error?.code === 'string') code = error.code;
			} catch {
				/* Untrusted non-JSON bodies are never echoed. */
			}
			throw new EvalsError(redactEvalsText(message, auth), response.status, redactEvalsText(code, auth));
		}
		if (raw) return text as T;
		try {
			const result = JSON.parse(text);
			if (!result || typeof result !== 'object' || Array.isArray(result)) throw new Error();
			return result as T;
		} catch {
			throw new EvalsError('Managed evaluation response is invalid', undefined, 'invalid_response');
		}
	}

	capabilities(): Promise<EvaluationCapabilities> {
		return this.request('GET', '/capabilities');
	}
	list(projectId?: string): Promise<{ evaluations: ManagedEvaluation[] }> {
		return this.request('GET', '/evaluations' + (projectId === undefined ? '' : '?' + new URLSearchParams({ projectId })));
	}
	get(evaluationId: string): Promise<{ evaluation: ManagedEvaluation; revisions: Array<{ revision: number; createdAt: string; specHash: string }> }> {
		return this.request('GET', `/evaluations/${id(evaluationId)}`);
	}
	create(spec: EvaluationSpec): Promise<{ evaluation: ManagedEvaluation }> {
		return this.request('POST', '/evaluations', spec);
	}
	revise(evaluationId: string, spec: EvaluationSpec, expectedRevision: number): Promise<{ evaluation: ManagedEvaluation }> {
		revision(expectedRevision, 'expected-revision');
		return this.request('POST', `/evaluations/${id(evaluationId)}/revisions`, { spec, expectedRevision });
	}
	run(evaluationId: string, options: EvaluationRunOptions): Promise<EvaluationRunResponse> {
		revision(options.revision);
		id(options.idempotencyKey);
		if (options.baselineRunId !== undefined) id(options.baselineRunId);
		return this.request('POST', `/evaluations/${id(evaluationId)}/runs`, options);
	}
	runs(evaluationId?: string): Promise<{ runs: EvaluationRun[] }> {
		return this.request('GET', '/runs' + (evaluationId === undefined ? '' : '?' + new URLSearchParams({ evaluationId })));
	}
	status(runId: string): Promise<EvaluationRunResponse> {
		return this.request('GET', `/runs/${id(runId)}`);
	}
	cancel(runId: string): Promise<EvaluationRunResponse> {
		return this.request('POST', `/runs/${id(runId)}/cancel`, {});
	}
	baseline(evaluationId: string, runId: string): Promise<{ evaluation: ManagedEvaluation }> {
		id(runId);
		return this.request('POST', `/evaluations/${id(evaluationId)}/baseline`, { runId });
	}
	report(runId: string, format: 'junit'): Promise<string>;
	report(runId: string, format?: 'json'): Promise<Record<string, unknown>>;
	report(runId: string, format: 'json' | 'junit'): Promise<string | Record<string, unknown>>;
	report(runId: string, format: 'json' | 'junit' = 'json'): Promise<string | Record<string, unknown>> {
		if (!['json', 'junit'].includes(format)) throw new Error('Report format must be json or junit');
		return this.request('GET', `/runs/${id(runId)}/report?format=${format}`, undefined, format === 'junit');
	}
	async compare(runId: string): Promise<EvaluationRunResponse> {
		const result = await this.status(runId);
		if (!result.run?.comparison || typeof result.run.comparison !== 'object') throw new EvalsError('Run has no server comparison; create a run with baselineRunId', undefined, 'comparison_unavailable');
		return result;
	}
	review(runId: string, review: EvaluationReview): Promise<EvaluationRunResponse> {
		revision(review.expectedReportRevision, 'expected-report-revision');
		id(review.caseId);
		id(review.scorerId);
		if (review.caseResultId !== undefined) id(review.caseResultId);
		if (!['pass', 'fail', 'abstain'].includes(review.status) || typeof review.reason !== 'string' || !review.reason.trim()) throw new Error('Review needs a pass, fail, or abstain status and a reason');
		return this.request('POST', `/runs/${id(runId)}/review`, review);
	}
	assist(instruction: string, spec: EvaluationSpec): Promise<{ message: string; spec?: EvaluationSpec }> {
		return this.request('POST', '/assistant', { instruction, spec });
	}
	async wait(runId: string, { timeout = 300_000, pollInterval = 1_000 }: EvaluationWaitOptions = {}): Promise<EvaluationRunResponse> {
		positive(timeout, 'wait-timeout');
		positive(pollInterval, 'poll-interval');
		const path = `/runs/${id(runId)}`;
		const deadline = performance.now() + timeout;
		for (;;) {
			const remaining = deadline - performance.now();
			if (remaining <= 0) throw new EvalsError('Wait timed out; the durable run can be resumed with status', undefined, 'timeout');
			const result = await this.request<EvaluationRunResponse>('GET', path, undefined, false, remaining);
			if (['completed', 'cancelled', 'error'].includes(result.run?.status)) return result;
			if (!['queued', 'running'].includes(result.run?.status)) throw new EvalsError('Managed run response is invalid', undefined, 'invalid_response');
			await new Promise((resolve) => setTimeout(resolve, Math.min(pollInterval, Math.max(0, deadline - performance.now()))));
		}
	}
}

/** CI gate: 0 pass, 1 fail, 2 incomplete, active, cancelled or error. */
export function evaluationGateExitCode(run: EvaluationRun): number {
	if (run?.status !== 'completed') return 2;
	return run.summary?.gate === 'pass' ? 0 : run.summary?.gate === 'fail' ? 1 : 2;
}

File: packages/client-python/src/rocketride/evals.py
"""Managed evaluations over bearer HTTP, independent of the offline runner.

Methods return the v1 response envelopes unchanged. Requests are never retried
or redirected; callers retain the idempotency key when resuming a run request.
"""

import asyncio
import http.client
import json
import math
import re
import socket
import threading
import time
from typing import Any, Callable, Dict, Optional, Tuple
from urllib.parse import quote, urlencode, urlsplit, urlunsplit


class EvalsError(Exception):
    """A managed API failure with a safe message, HTTP status and server code."""

    def __init__(self, message: str, *, status: Optional[int] = None, code: str = 'evals_error'):
        super().__init__(message)
        self.status = status
        self.code = code


def _redact(text: str, credential: str) -> str:
    if credential:
        text = text.replace(credential, '[REDACTED]')
    return re.sub(r'(?i)Bearer\s+[^\s"<>]+', 'Bearer [REDACTED]', text)


def _base_url(uri: str) -> str:
    """Validate the original endpoint before discarding its WebSocket path."""
    try:
        if not uri or any(c.isspace() or ord(c) < 32 for c in uri) or '\\' in uri:
            raise ValueError
        parsed = urlsplit(uri if '://' in uri else 'http://' + uri)
        if (
            parsed.scheme not in ('http', 'https', 'ws', 'wss')
            or not parsed.hostname
            or parsed.username is not None
            or parsed.password is not None
            or '?' in uri
            or '#' in uri
            or parsed.port == 0
            or parsed.path.rstrip('/') not in ('', '/task/service', '/evals/v1')
        ):
            raise ValueError
        scheme = {'ws': 'http', 'wss': 'https'}.get(parsed.scheme, parsed.scheme)
        return urlunsplit((scheme, parsed.netloc, '/evals/v1', '', ''))
    except ValueError:
        raise ValueError(
            'Managed evaluations require an HTTP(S) server origin without credentials, query, or fragment'
        ) from None


def _positive(value: float, name: str) -> None:
    if isinstance(value, bool) or not isinstance(value, (int, float)) or not math.isfinite(value) or value <= 0:
        raise ValueError(f'{name} must be a positive finite number')


def _revision(value: int, name: str = 'revision') -> None:
    if isinstance(value, bool) or not isinstance(value, int) or value < 1:
        raise ValueError(f'{name} must be a positive integer')


def _id(value: str) -> str:
    if not isinstance(value, str) or not value.strip() or value in ('.', '..'):
        raise ValueError('A non-empty object ID is required')
    return quote(value, safe='')


async def _http_request(
    method: str, url: str, headers: dict, body: Optional[bytes], timeout: float
) -> Tuple[int, bytes]:
    # http.client does not follow redirects, use environment proxies, or retry.
    # A daemon owns blocking DNS/socket work: a cancelled call must not hold
    # asyncio.run() open through its default executor's shutdown, or submit
    # a delayed POST when DNS eventually returns.
    parsed = urlsplit(url)
    connection_type = http.client.HTTPSConnection if parsed.scheme == 'https' else http.client.HTTPConnection
    connection = connection_type(parsed.hostname, parsed.port, timeout=timeout)
    loop = asyncio.get_running_loop()
    future = loop.create_future()
    cancelled = threading.Event()
    active_socket = None
    deadline = time.monotonic() + timeout

    def finish(value, error):
        if not future.done():
            if error is not None:
                future.set_exception(error)
            else:
                future.set_result(value)

    def send() -> None:
        nonlocal active_socket
        value, error = None, None
        try:
            connection.connect()
            active_socket = connection.sock
            if cancelled.is_set() or time.monotonic() >= deadline:
                raise TimeoutError
            # Explicitly disable implicit reconnection after this one connect.
            connection.auto_open = 0
            active_socket.settimeout(max(0.001, deadline - time.monotonic()))
            connection.request(method, urlunsplit(('', '', parsed.path, parsed.query, '')), body, headers)
            response = connection.getresponse()
            chunks = []
            while not response.isclosed():
                remaining = deadline - time.monotonic()
                if cancelled.is_set() or remaining <= 0:
                    raise TimeoutError
                # read1 bounds each read, so a slowly streaming body cannot
                # keep the socket alive beyond the request's total deadline.
                active_socket.settimeout(remaining)
                chunk = response.read1(65536)
                if not chunk:
                    break
                chunks.append(chunk)
            if response.length not in (None, 0):
                raise http.client.IncompleteRead(b'')
            value = response.status, b''.join(chunks)
        except Exception as caught:
            error = caught
        finally:
            connection.close()
            try:
                loop.call_soon_threadsafe(finish, value, error)
            except RuntimeError:
                pass  # The caller already cancelled and closed its loop.

    threading.Thread(target=send, name='rocketride-evals-http', daemon=True).start()
    try:
        return await future
    finally:
        cancelled.set()
        if active_socket is not None:
            try:
                active_socket.shutdown(socket.SHUT_RDWR)
            except OSError:
                pass
        connection.close()


class EvalsApi:
    """The ``client.evals`` namespace; no WebSocket connection is necessary.

    ``timeout`` is the per-request budget in seconds (default 30). ``wait``
    additionally bounds the complete polling period, including HTTP requests.
    """

    def __init__(self, connection: Callable[[], Tuple[str, str]], *, timeout: float = 30):
        _positive(timeout, 'timeout')
        self._connection = connection
        self.timeout = timeout

    async def _request(
        self, method: str, path: str, body: Any = None, *, raw: bool = False, timeout: Optional[float] = None
    ) -> Any:
        uri, credential = self._connection()
        url = _base_url(uri) + path
        if not credential or any(c.isspace() or ord(c) < 32 for c in credential):
            raise ValueError('A bearer credential is required for managed evaluations')
        budget = self.timeout if timeout is None else min(timeout, self.timeout)
        _positive(budget, 'timeout')
        headers = {'Authorization': f'Bearer {credential}', 'Accept': 'application/xml' if raw else 'application/json'}
        payload = None
        if body is not None:
            headers['Content-Type'] = 'application/json'
            try:
                payload = json.dumps(body, allow_nan=False).encode('utf-8')
            except (TypeError, ValueError):
                raise ValueError('Request body must be strict JSON') from None
        try:
            status, data = await asyncio.wait_for(_http_request(method, url, headers, payload, budget), budget)
        except asyncio.TimeoutError:
            raise EvalsError('Managed evaluation request timed out; no request was retried', code='timeout') from None
        except Exception:
            raise EvalsError(
                'Managed evaluation request failed; no request was retried', code='transport_error'
            ) from None
        if not 200 <= status < 300:
            message, code = f'Managed evaluation request failed (HTTP {status})', 'http_error'
            try:
                error = json.loads(data)['error']
                if isinstance(error.get('message'), str):
                    message = error['message']
                if isinstance(error.get('code'), str):
                    code = error['code']
            except (ValueError, TypeError, KeyError, AttributeError):
                pass
            raise EvalsError(_redact(message, credential), status=status, code=_redact(code, credential))
        try:
            if raw:
                return data.decode('utf-8')
            result = json.loads(data)
            if not isinstance(result, dict):
                raise ValueError
            return result
        except (ValueError, UnicodeError):
            raise EvalsError('Managed evaluation response is invalid', code='invalid_response') from None

    async def capabilities(self) -> Dict[str, Any]:
        return await self._request('GET', '/capabilities')

    async def list(self, *, project_id: Optional[str] = None) -> Dict[str, Any]:
        query = '?' + urlencode({'projectId': project_id}) if project_id is not None else ''
        return await self._request('GET', '/evaluations' + query)

    async def get(self, evaluation_id: str) -> Dict[str, Any]:
        return await self._request('GET', '/evaluations/' + _id(evaluation_id))

    async def create(self, spec: Dict[str, Any]) -> Dict[str, Any]:
        return await self._request('POST', '/evaluations', spec)

    async def revise(self, evaluation_id: str, spec: Dict[str, Any], *, expected_revision: int) -> Dict[str, Any]:
        _revision(expected_revision, 'expected-revision')
        return await self._request(
            'POST',
            f'/evaluations/{_id(evaluation_id)}/revisions',
            {'spec': spec, 'expectedRevision': expected_revision},
        )

    async def run(
        self, evaluation_id: str, *, revision: int, idempotency_key: str, baseline_run_id: Optional[str] = None
    ) -> Dict[str, Any]:
        _revision(revision)
        _id(idempotency_key)
        body = {'revision': revision, 'idempotencyKey': idempotency_key}
        if baseline_run_id is not None:
            _id(baseline_run_id)
            body['baselineRunId'] = baseline_run_id
        return await self._request('POST', f'/evaluations/{_id(evaluation_id)}/runs', body)

    async def runs(self, *, evaluation_id: Optional[str] = None) -> Dict[str, Any]:
        query = '?' + urlencode({'evaluationId': evaluation_id}) if evaluation_id is not None else ''
        return await self._request('GET', '/runs' + query)

    async def status(self, run_id: str) -> Dict[str, Any]:
        return await self._request('GET', '/runs/' + _id(run_id))

    async def cancel(self, run_id: str) -> Dict[str, Any]:
        return await self._request('POST', f'/runs/{_id(run_id)}/cancel', {})

    async def baseline(self, evaluation_id: str, run_id: str) -> Dict[str, Any]:
        _id(run_id)
        return await self._request('POST', f'/evaluations/{_id(evaluation_id)}/baseline', {'runId': run_id})

    async def report(self, run_id: str, *, format: str = 'json') -> Any:
        if format not in ('json', 'junit'):
            raise ValueError('Report format must be json or junit')
        return await self._request('GET', f'/runs/{_id(run_id)}/report?format={format}', raw=format == 'junit')

    async def compare(self, run_id: str) -> Dict[str, Any]:
        """Read the server's comparison against the run's recorded baseline."""
        result = await self.status(run_id)
        if not isinstance(result.get('run'), dict) or not isinstance(result['run'].get('comparison'), dict):
            raise EvalsError(
                'Run has no server comparison; create a run with baselineRunId', code='comparison_unavailable'
            )
        return result

    async def review(
        self,
        run_id: str,
        *,
        case_id: str,
        scorer_id: str,
        status: str,
        reason: str,
        expected_report_revision: int,
        case_result_id: Optional[str] = None,
    ) -> Dict[str, Any]:
        _revision(expected_report_revision, 'expected-report-revision')
        _id(case_id)
        _id(scorer_id)
        if status not in ('pass', 'fail', 'abstain') or not isinstance(reason, str) or not reason.strip():
            raise ValueError('Review needs a pass, fail, or abstain status and a reason')
        body = {
            'caseId': case_id,
            'scorerId': scorer_id,
            'status': status,
            'reason': reason,
            'expectedReportRevision': expected_report_revision,
        }
        if case_result_id is not None:
            _id(case_result_id)
            body['caseResultId'] = case_result_id
        return await self._request('POST', f'/runs/{_id(run_id)}/review', body)

    async def assist(self, instruction: str, spec: Dict[str, Any]) -> Dict[str, Any]:
        return await self._request('POST', '/assistant', {'instruction': instruction, 'spec': spec})

    async def wait(self, run_id: str, *, timeout: float = 300, poll_interval: float = 1) -> Dict[str, Any]:
        """Poll once immediately, then until terminal or deadline; never cancel."""
        _positive(timeout, 'wait-timeout')
        _positive(poll_interval, 'poll-interval')
        path = '/runs/' + _id(run_id)
        deadline = time.monotonic() + timeout
        while True:
            remaining = deadline - time.monotonic()
            if remaining <= 0:
                raise EvalsError('Wait timed out; the durable run can be resumed with status', code='timeout')
            result = await self._request('GET', path, timeout=remaining)
            run = result.get('run', {})
            if run.get('status') in ('completed', 'cancelled', 'error'):
                return result
            if run.get('status') not in ('queued', 'running'):
                raise EvalsError('Managed run response is invalid', code='invalid_response')
            await asyncio.sleep(min(poll_interval, max(0, deadline - time.monotonic())))


def gate_exit_code(run: Dict[str, Any]) -> int:
    """CI gate: 0 pass, 1 fail, 2 incomplete, active, cancelled or error."""
    if run.get('status') != 'completed':
        return 2
    return {'pass': 0, 'fail': 1}.get(run.get('summary', {}).get('gate'), 2)
